memflow-kmod/vmtools.c

```diff
@@ -292,20 +299,21 @@ static int memflow_vm_mem_mmap(struct file *file, struct vm_area_struct *vma)
 	}
 
 	if (ret || data->nr_pages != nr_pages)
-		goto put_pages;
+		goto free_pages;
 
 	data->wrapped_vma = NULL;
 	data->wrapped_task = NULL;
 
 	return 0;
 
-put_pages:
-	for (i = 0; i < data->nr_pages; i++)
-		RELEASE_PAGE(data->pages[i]);
 free_pages:
-	vfree(data->pages);
-	data->nr_pages = 0;
-	data->pages = NULL;
+	if (data->pages != NULL) {
+		for (i = 0; i < data->nr_pages; i++)
```

Comment:
This introduces an extra release bug on <6.5.0 kernels - if vmalloc of tmp_vmas fails, this branch will be executed, before get_user_pages_remote is invoked, leading to the refcount of pages going down.

I would revert the free flow to the old one, and just ifdef the `free_pages` label, because it is not needed on post 6.5

Reply:
Resolved in 0babac7